adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorTest.java

```diff
@@ -0,0 +1,146 @@
+
+        final CredentialsClient credentialsClient = mock(CredentialsClient.class);
+        final CredentialsObject credentialsOnRecord = new CredentialsObject(TEST_DEVICE, AUTH_ID, "hashed-password");
+        when(credentialsClient.get(anyString(), anyString())).thenReturn(Future.succeededFuture(credentialsOnRecord));
+
+        credentialsServiceClient = mock(HonoClient.class);
+        when(credentialsServiceClient.getOrCreateCredentialsClient(anyString()))
+                .thenReturn(Future.succeededFuture(credentialsClient));
+
+        authProvider = mock(HonoClientBasedAuthProvider.class);
+        authenticatorFactory = new AmqpAdapterSaslAuthenticatorFactory(authProvider, config);
+    }
+
+    /**
+     * Cleans up the fixture for this test suite.
+     */
+    @After
+    public void after() {
+        if (vertx != null) {
+            vertx.close();
+        }
+    }
+
+    /**
+     * Verifies that the SASL authenticator successfully authenticates a device with
+     * valid credentials (i.e the credentials supplied by the device matches those found in the
+     * devices' credentials record).
+     * 
+     * @param context The Vert.x context for running asynchronous tests.
+     */
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testAuthenticationWithValidCredentialsSucceeds(final TestContext context) {
+        final Async completionTracker = context.async();
+        final Device expectedDevice = mock(Device.class);
+
+        doAnswer(answer -> {
+            final Handler<AsyncResult<Device>> resultHandler = answer.getArgument(1);
+            resultHandler.handle(Future.succeededFuture(expectedDevice));
+            return null;
+        }).when(authProvider).authenticate(any(DeviceCredentials.class), any(Handler.class));
+
+        ProtonServer.create(vertx).saslAuthenticatorFactory(authenticatorFactory)
+                .connectHandler(serverConnection -> {
+                    // verify the expected device was recorded in the connection's attachments
+                    final Device actualDevice = serverConnection.attachments().get(Constants.KEY_CLIENT_DEVICE,
+                            Device.class);
+                    context.assertEquals(expectedDevice, actualDevice);
+
+                    completionTracker.complete();
+                }).listen(server -> {
+                    ProtonClient.create(vertx).connect("localhost", server.result().actualPort(), VALID_USERNAME,
+                            PASSWORD,
+                            clientConnection -> {
+                                // verify that credentials were successfully verified and the connection was successful.
+                                context.assertTrue(clientConnection.succeeded());
+                                clientConnection.result().disconnect();
+                                completionTracker.complete();
+                            });
+                });
+        completionTracker.awaitSuccess();
```

Comment:
are you sure that you need to start a real AMQP server and client in order to verify the authenticator's behavior?

Reply:
IMHO, having a client and server can trigger the SASL authentication process. and since the authenticator is used together with the proton server, i decided to test using a real server and client. WDYT?

Reply:
FMPOV a unit test should try to narrow down the scope of the test to a single class. Running a client and server with real networking involved sounds more like an integration test to me. Maybe you can move this to the IT module?